### ISSUE TITLE
Proposal: an easy-to-use pipeline package just for PK/PD

### ISSUE THREAD

**Issue by wlandau** [nlmixrdevelopment/nlmixr#447]
Folks like @billdenney have used `nlmixr` successfully with [`drake`](https://github.com/wlandau/drake), so I would like to chime in about how `nlmixr` could potentially partner with [`drake`](https://github.com/wlandau/drake)'s long-term successor, [`targets`](https://github.com/wlandau/targets).

Unlike [`drake`](https://github.com/wlandau/drake), [`targets`](https://github.com/wlandau/targets) makes it super straightforward to build extensions for specialized use cases. For example, I recently created a new R package called [`stantargets`](https://github.com/wlandau/stantargets), which combines [`targets`](https://github.com/wlandau/targets) with [`cmdstanr`](https://github.com/stan-dev/cmdstanr) to make it easy to write large Bayesian data analysis pipelines. Complicated workflows like [this one](https://github.com/wlandau/targets-stan/blob/main/_targets.R) condense down to [this one](https://wlandau.github.io/stantargets/articles/mcmc_rep.html) without loss of technical sophistication or computing power. The former requires users to think carefully about dynamic branching and file tracking, and the latter uses domain knowledge to abstract away these intimidating concepts. I though you might like to know because it instantiates a successful pattern that my colleagues and I are actively pursuing: develop one general-purpose pipeline tool first, and then build friendlier methodology-specific workflow frameworks on top of that. Other potential directions include stantargets-like packages for JAGS, Greta, Keras, Torch, etc.

For you, I see potential in a [`stantargets`](https://github.com/wlandau/stantargets)-like package tailored to `nlmixr`. I do not have the domain knowledge to develop it myself, but I would be happy to discuss ideas and review code if this appeals to you.

**Comment by billdenney** [comment]
@mattfidler, I'm moving into the world of `targets` now.  So, I'm looking into making `tar_nlmixr()` as a mapping to `nlmixr()`.  For that, I'm thinking about how to break `nlmixr()` into its component parts.  My best guess is that it would break into:

* model (function) to nlmixrUI object
* nlmixrUI object to compiled RxODE object
* Estimate the fit from the RxODE object and data combined

Perhaps the first two should be combined.  Also, this breakdown assumes that this set of steps each becomes an object that is readily cache-able.

Does that look right to you?  Any hints on how to break it down?

**Comment by mattfidler** [comment]
Hi @billdenney,

Thanks for the interest.  As mentioned in other issues I am going to allow the `nlmixr` functions to be used `RxODE`.  Because of this, the `ini` block is being moved to `lotri`. The `model` is being moved to `RxODE` and the interface to `saem` and `focei` will be split between `RxODE` and `nlmixr`.

Because of this, I think this requires careful thought.  If you have time maybe we can discuss this in a meeting to see the requirements and perhaps come up with a solution that will be backward and forward compatible.

**Comment by billdenney** [comment]
Hi @mattfidler, thanks for the chat just now to clarify the best ways to setup the targets.  As discussed, there will be 3 steps:

* Model object creation
* Data preparation
* Estimation

For the model object creation:
* Drop the srcref (`attr(f, "srcref") <- NULL`) to make the model not depend on comments
* Check into environment dependence

For data preparation:
* Determine required covariate columns from the model object
* Select normal columns (e.g. `ID`, `DV`, etc.) and the covariate columns

If the model object and/or the data differ, estimate.

**Comment by wlandau** [comment]
> Is the method I'm using with substitute() the right way to set the targets up with tar_target_raw()?

Yes, you spend most of the target factory using `substitute()` to create the commands from user input, and you delegate the actual work and domain knowledge to other functions. Looks great.

> Is returning a list from tar_nlmixr() made up of the targets to run the right way to create the DSL-specific tarchetype?

Yes, I recommend returning a list of target objects. Target lists can be arbitrarily nested in any tree structure. As long as each leaf node is a target object, the `targets` package will accept it and flatten out the list automatically.

> I want to be sure that my data_simple and object_simple targets do not conflict with any other names. 

I think you will be fine if different values of `name` produce different sets of target names. Conflicts are indeed possible between `tar_nlmixr(simple)` and `tar_nlmixr(data_simple)`, but this seems unlikely in practice, and `targets` will error out before any work is wasted. I use a similar naming strategy in `stantargets` and `jagstargets`, and it seems to work fine.

Aside: I generally find it more convenient to put `name` first in the target names, e.g. `simple_data` and `simple_object`. 

> I'm wanting to do this because there will be times where potentially hundreds of models (i.e. calls to tar_nlmixr()) could be created in a single target set. 

Might be worth considering about dynamic branching. Not saying it's the right thing to do, I really have no idea, but this is the right time to think about it. If you need small `tar_visnetwork()` graphs under heavy load, or if the choices of models come from some upstream part of the computation, then those factors could motivate dynamic branching. It may also help reduce potential name conflicts because different models would be given unique hash suffixes instead of multiple user-supplied names.

> I want to give ... arguments to tar_nlmixr() so that they can be passed to nlmixr::nlmixr() in the last step, but I don't see a way to do this. 

Inside `tar_nlmixr()`, you can capture those arguments using `list(...)`. Then you can add more arguments to that argument list, like `object`, and then write the command as `do.call(nlmixr::nlmixr, args = your_argument_list)`. 

But wherever you can, I recommend formal arguments. The job of a target factory is to create a waterproof abstraction, whereas `...` and `.target_args=list()` are leaky. I realize `nlmixr()` does take `...`, so maybe we're stuck with that. But I do recommend that all the `targets`-specific arguments have formal names as in other target factories like `stantargets::tar_stan_mcmc()`.

**Comment by billdenney** [comment]
@wlandau , Thanks for the responses; they are very helpful.  The handling of `...` will take a bit more thought, and I appreciate your thoughts on it.

I'll switch the space in the target name to something likely to be unique but better behaved and allow for the fact that `targets` will take care of conflicts for me.

@mattfidler , I think that this is likely a better fit to start in `nlmixrExtra`, and that it may eventually move to a new package (`nlmixrtargets`?).  Does that make sense to you?  Would you prefer that it immediately move to `nlmixrtargets`?

**Comment by wlandau** [comment]
I would recommend a new nlmixrtargets package if there is a chance you will create more than 1 or 2 target factories.

**Comment by billdenney** [comment]
@wlandau That makes sense.  @mattfidler , if you agree, can you please create the repo and give me access?

**Comment by wlandau** [comment]
A couple other bits of logistics if you are interested:

* Feel free to borrow the R Targetopia badge for the README of `nlmixrtargets`: [![R
Targetopia](https://img.shields.io/badge/R_Targetopia-member-blue?style=flat&labelColor=gray)](https://wlandau.github.io/targetopia/)
* If you have a hex sticker and abstract, I would love to host it at https://wlandau.github.io/targetopia/packages.html. PRs welcome at https://github.com/wlandau/targetopia.
